stl/inc/flat_map

```diff
@@ -0,0 +1,488 @@
+    }
+
+    template <class _Self_t>
+    _NODISCARD auto end(this _Self_t&& _Self) noexcept {
+        return _Maybe_const_iterator_t<_Self_t&&> { _Self._Data.keys.cend(), _Self._Data.values.end() };
+    }
+
+    template <class _Self_t>
+    _NODISCARD auto rbegin(this _Self_t&& _Self) noexcept {
+        return _STD make_reverse_iterator(_Self.end());
+    }
+
+    template <class _Self_t>
+    _NODISCARD auto rend(this _Self_t&& _Self) noexcept {
+        return _STD make_reverse_iterator(_Self.begin());
+    }
+
+    _NODISCARD const_iterator cbegin() const noexcept {
+        return const_iterator {_Data.keys.cbegin(), _Data.values.cbegin()};
+    }
+
+    _NODISCARD const_iterator cend() const noexcept {
+        return const_iterator {_Data.keys.cend(), _Data.values.cend()};
+    }
+
+    _NODISCARD const_reverse_iterator crbegin() const noexcept {
+        return _STD make_reverse_iterator(cend());
+    }
+
+    _NODISCARD const_reverse_iterator crend() const noexcept {
+        return _STD make_reverse_iterator(cbegin());
+    }
+
+    // [container.reqmts] clear
+    void clear() noexcept {
+        _Data.keys.clear();
+        _Data.values.clear();
+    }
+
+    // [flatmap.capacity] Capacity
+    _NODISCARD_EMPTY_MEMBER bool empty() const noexcept {
+        return _Data.keys.empty();
+    }
+
+    _NODISCARD size_type size() const noexcept {
+        return _Data.keys.size();
+    }
+
+    _NODISCARD size_type max_size() const noexcept {
+        return _STD min(_Data.keys.max_size(), _Data.values.max_size());
+    }
+
+    // [flatmap.access] Access
+    _NODISCARD mapped_type& operator[](const key_type& _Key); // TODO: implement
+    _NODISCARD mapped_type& operator[](key_type&& _Key); // TODO: implement
+    template <class _C>
+    _NODISCARD mapped_type& operator[](_C&& _Key)
+        requires _Compare_is_transparent<key_compare>; // TODO: implement
+
+    template <class _Self_t>
+    _NODISCARD auto&& at(this _Self_t&& _Self, const key_type& _Key); // TODO: implement
+    template <class _Self_t, class _C>
+    _NODISCARD auto&& at(this _Self_t&& _Self, const _C& _Key)
+        requires _Compare_is_transparent<key_compare>; // TODO: implement
+
+    // [flatmap.modifiers] Modifiers
+    template <class... _Args_t>
+    _STD pair<iterator, bool> emplace(_Args_t&&... _Args) requires _STD is_constructible_v<value_type, _Args_t...> {
+        value_type _Val(_STD forward<_Args_t>(_Args)...);
+        iterator _It = lower_bound(_Data.keys.begin(), _Data.keys.end(), _Val.first, _Key_compare);
+
+        if(_It != end() && *(_It._Key_it) == _Val.first) {
+            // Already exists
+            return _STD make_pair(_It, false);
+        }
+        else{
+            // Need to insert
+            auto _Index = _STD distance(begin(), _It);
+            _Data.keys.insert(_It._Key_it, _STD move(_Val.first));
+            _Data.values.insert(_It._Mapped_it, _STD move(_Val.second));

```

Comment:
notice that the two `insert` can potentially fail and break invariant here. does `emplace` require that `flat_map` should be "in a valid but unspecified state" in this case?

Reply:
I wasn't thinking of the invariants but yes, I think we need to maintain them even if `emplace` fails. [flatmap.overview] says:

> If any member function in [flatmap.defn] exits via an exception the invariants are restored.

Any suggestions on how to do this? How about:
```
try{
  _Data.keys.insert(_It._Key_it, _STD move(_Val.first));
  _Data.values.insert(_It._Mapped_it, _STD move(_Val.second));
}
catch(...){
  clear();
}
```

Reply:
We prefer scope-guard to try-catch. You can have a look at what is done for `flat_set`:
https://github.com/microsoft/STL/blob/1f4920713ecfed75de12eeca8375e4cab4e1ba56/stl/inc/flat_set#L22-L30